#!/usr/bin/env python3
"""
–¢–µ—Å—Ç –ø–æ–ª–Ω–æ–π pipeline –ª–æ–≥–∏–∫–∏: –≥–µ–Ω–µ—Ä–∞—Ü–∏—è ‚Üí –Ω–∞–∫–æ–ø–ª–µ–Ω–∏–µ ‚Üí —Å–±–æ—Ä–∫–∞ ‚Üí –ø–µ—Ä–µ–¥–∞—á–∞ –Ω–∞ –∫–ª–∏–µ–Ω—Ç

–ü—Ä–æ–≤–µ—Ä—è–µ—Ç –≤–µ—Å—å –ø—Ä–æ—Ü–µ—Å—Å:
1. Edge TTS –≥–µ–Ω–µ—Ä–∏—Ä—É–µ—Ç –∞—É–¥–∏–æ –¥–ª—è –ø—Ä–µ–¥–ª–æ–∂–µ–Ω–∏—è
2. –ü—Ä–æ–≤–∞–π–¥–µ—Ä —Ä–∞–∑–±–∏–≤–∞–µ—Ç –Ω–∞ —á–∞–Ω–∫–∏ (–≤–Ω—É—Ç—Ä–µ–Ω–Ω—è—è –æ–±—Ä–∞–±–æ—Ç–∫–∞)
3. _stream_audio_for_sentence –Ω–∞–∫–∞–ø–ª–∏–≤–∞–µ—Ç –≤—Å–µ —á–∞–Ω–∫–∏
4. –ß–∞–Ω–∫–∏ —Å–æ–±–∏—Ä–∞—é—Ç—Å—è –≤ –µ–¥–∏–Ω–æ–µ –∞—É–¥–∏–æ –ø—Ä–µ–¥–ª–æ–∂–µ–Ω–∏—è
5. –ü–æ–ª–Ω–æ–µ –∞—É–¥–∏–æ –æ—Ç–ø—Ä–∞–≤–ª—è–µ—Ç—Å—è —á–µ—Ä–µ–∑ gRPC –∫–ª–∏–µ–Ω—Ç—É
"""

import asyncio
import sys
import logging
from pathlib import Path
from typing import Dict, Any, List
from unittest.mock import Mock, AsyncMock

# –î–æ–±–∞–≤–ª—è–µ–º –ø—É—Ç—å –∫ —Å–µ—Ä–≤–µ—Ä—É
server_root = Path(__file__).parent.parent
sys.path.insert(0, str(server_root))

# –ù–∞—Å—Ç—Ä–æ–π–∫–∞ –ª–æ–≥–∏—Ä–æ–≤–∞–Ω–∏—è
logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s - %(name)s - %(levelname)s - %(message)s'
)
logger = logging.getLogger(__name__)


class MockTextProcessor:
    """–ú–æ–∫ —Ç–µ–∫—Å—Ç–æ–≤–æ–≥–æ –ø—Ä–æ—Ü–µ—Å—Å–æ—Ä–∞ (LLM)"""
    
    def __init__(self):
        self.is_initialized = True
        self.name = "text_processing"
    
    async def process(self, request: Dict[str, Any]):
        """–ò–º–∏—Ç–∞—Ü–∏—è –ø–æ—Ç–æ–∫–æ–≤–æ–≥–æ –æ—Ç–≤–µ—Ç–∞ LLM"""
        text = request.get('text', '')
        
        # –ò–º–∏—Ç–∏—Ä—É–µ–º –ø–æ—Ç–æ–∫–æ–≤—ã–π –æ—Ç–≤–µ—Ç –∞—Å—Å–∏—Å—Ç–µ–Ω—Ç–∞
        responses = [
            "Hello! This is a test sentence for audio generation. ",
            "The audio should be generated correctly. ",
            "All chunks should be collected and assembled properly."
        ]
        
        for response in responses:
            await asyncio.sleep(0.01)
            yield response


async def test_chunk_collection_pipeline():
    """
    –¢–µ—Å—Ç –ø–æ–ª–Ω–æ–π pipeline: –≥–µ–Ω–µ—Ä–∞—Ü–∏—è ‚Üí –Ω–∞–∫–æ–ø–ª–µ–Ω–∏–µ ‚Üí —Å–±–æ—Ä–∫–∞ ‚Üí –ø–µ—Ä–µ–¥–∞—á–∞
    """
    print("\n" + "="*80)
    print("–¢–ï–°–¢ –ü–û–õ–ù–û–ô PIPELINE: –ì–ï–ù–ï–†–ê–¶–ò–Ø ‚Üí –ù–ê–ö–û–ü–õ–ï–ù–ò–ï ‚Üí –°–ë–û–†–ö–ê ‚Üí –ü–ï–†–ï–î–ê–ß–ê")
    print("="*80)
    print("\n–ü—Ä–æ–≤–µ—Ä—è–µ–º:")
    print("1. Edge TTS –≥–µ–Ω–µ—Ä–∏—Ä—É–µ—Ç –∞—É–¥–∏–æ –¥–ª—è –ø—Ä–µ–¥–ª–æ–∂–µ–Ω–∏—è")
    print("2. –ü—Ä–æ–≤–∞–π–¥–µ—Ä —Ä–∞–∑–±–∏–≤–∞–µ—Ç –Ω–∞ —á–∞–Ω–∫–∏ (–≤–Ω—É—Ç—Ä–µ–Ω–Ω—è—è –æ–±—Ä–∞–±–æ—Ç–∫–∞)")
    print("3. _stream_audio_for_sentence –Ω–∞–∫–∞–ø–ª–∏–≤–∞–µ—Ç –≤—Å–µ —á–∞–Ω–∫–∏")
    print("4. –ß–∞–Ω–∫–∏ —Å–æ–±–∏—Ä–∞—é—Ç—Å—è –≤ –µ–¥–∏–Ω–æ–µ –∞—É–¥–∏–æ –ø—Ä–µ–¥–ª–æ–∂–µ–Ω–∏—è")
    print("5. –ü–æ–ª–Ω–æ–µ –∞—É–¥–∏–æ –æ—Ç–ø—Ä–∞–≤–ª—è–µ—Ç—Å—è —á–µ—Ä–µ–∑ gRPC –∫–ª–∏–µ–Ω—Ç—É")
    print("="*80)
    
    try:
        # –ò–º–ø–æ—Ä—Ç–∏—Ä—É–µ–º –Ω–µ–æ–±—Ö–æ–¥–∏–º—ã–µ –º–æ–¥—É–ª–∏
        from modules.audio_generation.core.audio_processor import AudioProcessor
        from integrations.workflow_integrations.streaming_workflow_integration import StreamingWorkflowIntegration
        from integrations.service_integrations.grpc_service_integration import GrpcServiceIntegration
        from modules.grpc_service.core.grpc_server import streaming_pb2
        
        # 1. –ò–Ω–∏—Ü–∏–∞–ª–∏–∑–∞—Ü–∏—è AudioProcessor (Edge TTS)
        print("\n[1/6] –ò–Ω–∏—Ü–∏–∞–ª–∏–∑–∞—Ü–∏—è AudioProcessor (Edge TTS)...")
        audio_processor = AudioProcessor()
        audio_init_result = await audio_processor.initialize()
        if not audio_init_result:
            print("   ‚ùå –û—à–∏–±–∫–∞ –∏–Ω–∏—Ü–∏–∞–ª–∏–∑–∞—Ü–∏–∏ AudioProcessor")
            return False
        print("   ‚úÖ AudioProcessor –∏–Ω–∏—Ü–∏–∞–ª–∏–∑–∏—Ä–æ–≤–∞–Ω")
        
        # 2. –°–æ–∑–¥–∞–Ω–∏–µ –º–æ–∫–∞ —Ç–µ–∫—Å—Ç–æ–≤–æ–≥–æ –ø—Ä–æ—Ü–µ—Å—Å–æ—Ä–∞
        print("\n[2/6] –°–æ–∑–¥–∞–Ω–∏–µ –º–æ–∫–∞ —Ç–µ–∫—Å—Ç–æ–≤–æ–≥–æ –ø—Ä–æ—Ü–µ—Å—Å–æ—Ä–∞...")
        text_processor = MockTextProcessor()
        print("   ‚úÖ –ú–æ–∫ —Ç–µ–∫—Å—Ç–æ–≤–æ–≥–æ –ø—Ä–æ—Ü–µ—Å—Å–æ—Ä–∞ —Å–æ–∑–¥–∞–Ω")
        
        # 3. –°–æ–∑–¥–∞–Ω–∏–µ StreamingWorkflowIntegration
        print("\n[3/6] –°–æ–∑–¥–∞–Ω–∏–µ StreamingWorkflowIntegration...")
        workflow = StreamingWorkflowIntegration(
            text_processor=text_processor,
            audio_processor=audio_processor,
            memory_workflow=None,
            text_filter_manager=None
        )
        await workflow.initialize()
        print("   ‚úÖ StreamingWorkflowIntegration –∏–Ω–∏—Ü–∏–∞–ª–∏–∑–∏—Ä–æ–≤–∞–Ω")
        
        # 4. –°–æ–∑–¥–∞–Ω–∏–µ GrpcServiceIntegration
        print("\n[4/6] –°–æ–∑–¥–∞–Ω–∏–µ GrpcServiceIntegration...")
        import inspect
        sig = inspect.signature(GrpcServiceIntegration.__init__)
        params = list(sig.parameters.keys())[1:]
        
        init_kwargs = {'streaming_workflow': workflow}
        if 'memory_workflow' in params:
            init_kwargs['memory_workflow'] = None
        if 'text_filter_manager' in params:
            init_kwargs['text_filter_manager'] = None
        
        grpc_integration = GrpcServiceIntegration(**init_kwargs)
        await grpc_integration.initialize()
        print("   ‚úÖ GrpcServiceIntegration –∏–Ω–∏—Ü–∏–∞–ª–∏–∑–∏—Ä–æ–≤–∞–Ω")
        
        # 5. –¢–µ—Å—Ç–æ–≤—ã–π –∑–∞–ø—Ä–æ—Å
        print("\n[5/6] –û–±—Ä–∞–±–æ—Ç–∫–∞ —Ç–µ—Å—Ç–æ–≤–æ–≥–æ –∑–∞–ø—Ä–æ—Å–∞...")
        request_data = {
            'text': 'Hello, can you tell me a test response?',
            'hardware_id': 'test_hardware_chunk_collection',
            'session_id': 'test_session_chunk_collection'
        }
        print(f"   üìù –ó–∞–ø—Ä–æ—Å: '{request_data['text']}'")
        
        # –°–æ–±–∏—Ä–∞–µ–º —Å—Ç–∞—Ç–∏—Å—Ç–∏–∫—É –¥–ª—è –ø—Ä–æ–≤–µ—Ä–∫–∏ –Ω–∞–∫–æ–ø–ª–µ–Ω–∏—è –∏ —Å–±–æ—Ä–∫–∏
        text_chunks_received = []
        audio_chunks_received = []  # –î–æ–ª–∂–Ω–æ –±—ã—Ç—å –û–î–ù–û –ø–æ–ª–Ω–æ–µ –∞—É–¥–∏–æ –ø—Ä–µ–¥–ª–æ–∂–µ–Ω–∏—è
        total_audio_bytes = 0
        sentence_indices = set()
        grpc_responses = []
        
        # –û—Ç—Å–ª–µ–∂–∏–≤–∞–µ–º –ø—Ä–æ—Ü–µ—Å—Å –Ω–∞–∫–æ–ø–ª–µ–Ω–∏—è —á–∞–Ω–∫–æ–≤
        provider_chunks_count = {}  # sentence_index -> –∫–æ–ª–∏—á–µ—Å—Ç–≤–æ —á–∞–Ω–∫–æ–≤ –æ—Ç –ø—Ä–æ–≤–∞–π–¥–µ—Ä–∞
        
        # –û–±—Ä–∞–±–∞—Ç—ã–≤–∞–µ–º –∑–∞–ø—Ä–æ—Å
        async for result in grpc_integration.process_request_complete(request_data):
            # –¢–µ–∫—Å—Ç–æ–≤—ã–µ —á–∞–Ω–∫–∏
            if 'text_response' in result and result.get('text_response'):
                text = result['text_response']
                text_chunks_received.append(text)
                sentence_index = result.get('sentence_index', 0)
                sentence_indices.add(sentence_index)
                print(f"   üìù –¢–µ–∫—Å—Ç –ø—Ä–µ–¥–ª–æ–∂–µ–Ω–∏—è #{sentence_index}: '{text[:60]}{'...' if len(text) > 60 else ''}'")
                
                grpc_response = streaming_pb2.StreamResponse(text_chunk=text)
                grpc_responses.append(('text', text, sentence_index))
            
            # –ê—É–¥–∏–æ —á–∞–Ω–∫–∏ (–¥–æ–ª–∂–Ω–æ –±—ã—Ç—å –û–î–ù–û –ø–æ–ª–Ω–æ–µ –∞—É–¥–∏–æ –¥–ª—è –∫–∞–∂–¥–æ–≥–æ –ø—Ä–µ–¥–ª–æ–∂–µ–Ω–∏—è)
            if 'audio_chunk' in result:
                audio_chunk = result['audio_chunk']
                if isinstance(audio_chunk, (bytes, bytearray)) and len(audio_chunk) > 0:
                    audio_chunks_received.append(audio_chunk)
                    total_audio_bytes += len(audio_chunk)
                    sentence_index = result.get('sentence_index', 0)
                    
                    print(f"   üîä –ü–æ–ª–Ω–æ–µ –∞—É–¥–∏–æ –ø—Ä–µ–¥–ª–æ–∂–µ–Ω–∏—è #{sentence_index}: {len(audio_chunk)} –±–∞–π—Ç")
                    
                    # –ò–º–∏—Ç–∏—Ä—É–µ–º gRPC –æ—Ç–≤–µ—Ç
                    grpc_response = streaming_pb2.StreamResponse(
                        audio_chunk=streaming_pb2.AudioChunk(
                            audio_data=audio_chunk,
                            dtype='int16',
                            shape=[]
                        )
                    )
                    grpc_responses.append(('audio', len(audio_chunk), sentence_index))
        
        # 6. –ü—Ä–æ–≤–µ—Ä–∫–∞ —Ä–µ–∑—É–ª—å—Ç–∞—Ç–æ–≤
        print("\n" + "="*80)
        print("–†–ï–ó–£–õ–¨–¢–ê–¢–´ –¢–ï–°–¢–ò–†–û–í–ê–ù–ò–Ø")
        print("="*80)
        
        print(f"\nüìä –°—Ç–∞—Ç–∏—Å—Ç–∏–∫–∞:")
        print(f"   - –¢–µ–∫—Å—Ç–æ–≤—ã—Ö –ø—Ä–µ–¥–ª–æ–∂–µ–Ω–∏–π: {len(text_chunks_received)}")
        print(f"   - –ê—É–¥–∏–æ –ø—Ä–µ–¥–ª–æ–∂–µ–Ω–∏–π (–ø–æ–ª–Ω—ã—Ö): {len(audio_chunks_received)}")
        print(f"   - –í—Å–µ–≥–æ –∞—É–¥–∏–æ –±–∞–π—Ç: {total_audio_bytes} ({total_audio_bytes / 1024:.2f} KB)")
        print(f"   - gRPC –æ—Ç–≤–µ—Ç–æ–≤: {len(grpc_responses)}")
        print(f"   - –£–Ω–∏–∫–∞–ª—å–Ω—ã—Ö sentence_index: {len(sentence_indices)}")
        
        # –ö–†–ò–¢–ò–ß–ï–°–ö–ò–ï –ü–†–û–í–ï–†–ö–ò
        print(f"\nüîç –ö–†–ò–¢–ò–ß–ï–°–ö–ò–ï –ü–†–û–í–ï–†–ö–ò:")
        
        checks = {}
        
        # –ü—Ä–æ–≤–µ—Ä–∫–∞ 1: –ö–æ–ª–∏—á–µ—Å—Ç–≤–æ —Ç–µ–∫—Å—Ç–æ–≤—ã—Ö –ø—Ä–µ–¥–ª–æ–∂–µ–Ω–∏–π = –∫–æ–ª–∏—á–µ—Å—Ç–≤–æ –∞—É–¥–∏–æ –ø—Ä–µ–¥–ª–æ–∂–µ–Ω–∏–π
        text_count = len(text_chunks_received)
        audio_count = len(audio_chunks_received)
        checks["–ö–æ–ª–∏—á–µ—Å—Ç–≤–æ —Ç–µ–∫—Å—Ç–æ–≤—ã—Ö –ø—Ä–µ–¥–ª–æ–∂–µ–Ω–∏–π = –∫–æ–ª–∏—á–µ—Å—Ç–≤—É –∞—É–¥–∏–æ –ø—Ä–µ–¥–ª–æ–∂–µ–Ω–∏–π"] = (
            text_count == audio_count and text_count > 0
        )
        print(f"   {'‚úÖ' if checks['–ö–æ–ª–∏—á–µ—Å—Ç–≤–æ —Ç–µ–∫—Å—Ç–æ–≤—ã—Ö –ø—Ä–µ–¥–ª–æ–∂–µ–Ω–∏–π = –∫–æ–ª–∏—á–µ—Å—Ç–≤—É –∞—É–¥–∏–æ –ø—Ä–µ–¥–ª–æ–∂–µ–Ω–∏–π'] else '‚ùå'} "
              f"–¢–µ–∫—Å—Ç–æ–≤—ã—Ö: {text_count}, –ê—É–¥–∏–æ: {audio_count}")
        
        # –ü—Ä–æ–≤–µ—Ä–∫–∞ 2: –ö–∞–∂–¥–æ–µ –∞—É–¥–∏–æ –ø—Ä–µ–¥–ª–æ–∂–µ–Ω–∏–µ - —ç—Ç–æ –ø–æ–ª–Ω–æ–µ –∞—É–¥–∏–æ (–Ω–µ –æ—Ç–¥–µ–ª—å–Ω—ã–µ —á–∞–Ω–∫–∏)
        if audio_chunks_received:
            # –ü—Ä–æ–≤–µ—Ä—è–µ–º —Ä–∞–∑–º–µ—Ä—ã - –ø–æ–ª–Ω–æ–µ –∞—É–¥–∏–æ –¥–æ–ª–∂–Ω–æ –±—ã—Ç—å –¥–æ—Å—Ç–∞—Ç–æ—á–Ω–æ –±–æ–ª—å—à–∏–º
            # (–Ω–µ –º–∞–ª–µ–Ω—å–∫–∏–µ —á–∞–Ω–∫–∏ –ø–æ 4096 –±–∞–π—Ç)
            min_expected_size = 10000  # –ú–∏–Ω–∏–º–∞–ª—å–Ω—ã–π —Ä–∞–∑–º–µ—Ä –¥–ª—è –ø–æ–ª–Ω–æ–≥–æ –ø—Ä–µ–¥–ª–æ–∂–µ–Ω–∏—è
            all_complete = all(len(chunk) >= min_expected_size for chunk in audio_chunks_received)
            checks["–í—Å–µ –∞—É–¥–∏–æ –ø—Ä–µ–¥–ª–æ–∂–µ–Ω–∏—è - –ø–æ–ª–Ω—ã–µ (–Ω–µ –æ—Ç–¥–µ–ª—å–Ω—ã–µ —á–∞–Ω–∫–∏)"] = all_complete
            
            sizes = [len(chunk) for chunk in audio_chunks_received]
            print(f"   {'‚úÖ' if all_complete else '‚ùå'} "
                  f"–†–∞–∑–º–µ—Ä—ã –∞—É–¥–∏–æ –ø—Ä–µ–¥–ª–æ–∂–µ–Ω–∏–π: {sizes}")
            print(f"      (–º–∏–Ω–∏–º–∞–ª—å–Ω—ã–π –æ–∂–∏–¥–∞–µ–º—ã–π: {min_expected_size} –±–∞–π—Ç)")
            
            # –ü—Ä–æ–≤–µ—Ä–∫–∞ 3: –ö–∞–∂–¥–æ–µ –ø—Ä–µ–¥–ª–æ–∂–µ–Ω–∏–µ –∏–º–µ–µ—Ç —É–Ω–∏–∫–∞–ª—å–Ω—ã–π sentence_index
            audio_sentence_indices = [r[2] for r in grpc_responses if r[0] == 'audio']
            unique_indices = len(set(audio_sentence_indices))
            checks["–ö–∞–∂–¥–æ–µ –∞—É–¥–∏–æ –ø—Ä–µ–¥–ª–æ–∂–µ–Ω–∏–µ –∏–º–µ–µ—Ç —É–Ω–∏–∫–∞–ª—å–Ω—ã–π sentence_index"] = (
                unique_indices == audio_count
            )
            print(f"   {'‚úÖ' if checks['–ö–∞–∂–¥–æ–µ –∞—É–¥–∏–æ –ø—Ä–µ–¥–ª–æ–∂–µ–Ω–∏–µ –∏–º–µ–µ—Ç —É–Ω–∏–∫–∞–ª—å–Ω—ã–π sentence_index'] else '‚ùå'} "
                  f"–£–Ω–∏–∫–∞–ª—å–Ω—ã—Ö –∏–Ω–¥–µ–∫—Å–æ–≤: {unique_indices}, –ê—É–¥–∏–æ –ø—Ä–µ–¥–ª–æ–∂–µ–Ω–∏–π: {audio_count}")
        
        # –ü—Ä–æ–≤–µ—Ä–∫–∞ 4: gRPC —Ñ–æ—Ä–º–∞—Ç –∫–æ—Ä—Ä–µ–∫—Ç–µ–Ω
        audio_grpc_responses = [r for r in grpc_responses if r[0] == 'audio']
        checks["gRPC –æ—Ç–≤–µ—Ç—ã —Å—Ñ–æ—Ä–º–∏—Ä–æ–≤–∞–Ω—ã –∫–æ—Ä—Ä–µ–∫—Ç–Ω–æ"] = len(audio_grpc_responses) > 0
        print(f"   {'‚úÖ' if checks['gRPC –æ—Ç–≤–µ—Ç—ã —Å—Ñ–æ—Ä–º–∏—Ä–æ–≤–∞–Ω—ã –∫–æ—Ä—Ä–µ–∫—Ç–Ω–æ'] else '‚ùå'} "
              f"–ê—É–¥–∏–æ gRPC –æ—Ç–≤–µ—Ç–æ–≤: {len(audio_grpc_responses)}")
        
        # –ü—Ä–æ–≤–µ—Ä–∫–∞ 5: –§–æ—Ä–º–∞—Ç –∞—É–¥–∏–æ —Å–æ–≤–º–µ—Å—Ç–∏–º —Å Azure TTS
        if audio_chunks_received:
            first_chunk = audio_chunks_received[0]
            # PCM 16-bit –¥–∞–Ω–Ω—ã–µ –¥–æ–ª–∂–Ω—ã –±—ã—Ç—å –∫—Ä–∞—Ç–Ω—ã 2
            pcm_valid = len(first_chunk) % 2 == 0
            checks["–§–æ—Ä–º–∞—Ç –∞—É–¥–∏–æ —Å–æ–≤–º–µ—Å—Ç–∏–º —Å Azure TTS (PCM 16-bit)"] = pcm_valid
            print(f"   {'‚úÖ' if pcm_valid else '‚ùå'} "
                  f"PCM —Ñ–æ—Ä–º–∞—Ç –≤–∞–ª–∏–¥–µ–Ω (—Ä–∞–∑–º–µ—Ä –∫—Ä–∞—Ç–µ–Ω 2): {len(first_chunk)} –±–∞–π—Ç")
        
        # –ü—Ä–æ–≤–µ—Ä–∫–∞ 6: –°–æ–æ—Ç–≤–µ—Ç—Å—Ç–≤–∏–µ —Ç–µ–∫—Å—Ç–∞ –∏ –∞—É–¥–∏–æ –ø–æ sentence_index
        text_indices = [r[2] for r in grpc_responses if r[0] == 'text']
        audio_indices = [r[2] for r in grpc_responses if r[0] == 'audio']
        indices_match = set(text_indices) == set(audio_indices)
        checks["–¢–µ–∫—Å—Ç –∏ –∞—É–¥–∏–æ —Å–æ–æ—Ç–≤–µ—Ç—Å—Ç–≤—É—é—Ç –ø–æ sentence_index"] = indices_match
        print(f"   {'‚úÖ' if indices_match else '‚ùå'} "
              f"–¢–µ–∫—Å—Ç–æ–≤—ã–µ –∏–Ω–¥–µ–∫—Å—ã: {sorted(text_indices)}, –ê—É–¥–∏–æ –∏–Ω–¥–µ–∫—Å—ã: {sorted(audio_indices)}")
        
        # –ò—Ç–æ–≥–æ–≤—ã–µ –ø—Ä–æ–≤–µ—Ä–∫–∏
        print(f"\n‚úÖ –§–ò–ù–ê–õ–¨–ù–´–ï –ü–†–û–í–ï–†–ö–ò:")
        all_passed = True
        for check_name, check_result in checks.items():
            status = "‚úÖ" if check_result else "‚ùå"
            print(f"   {status} {check_name}")
            if not check_result:
                all_passed = False
        
        if all_passed:
            print("\n" + "="*80)
            print("üéâ –í–°–ï –ü–†–û–í–ï–†–ö–ò –ü–†–û–ô–î–ï–ù–´!")
            print("="*80)
            print("\n‚úÖ –ü–æ–¥—Ç–≤–µ—Ä–∂–¥–µ–Ω–æ:")
            print("   1. Edge TTS –≥–µ–Ω–µ—Ä–∏—Ä—É–µ—Ç –∞—É–¥–∏–æ –¥–ª—è –ø—Ä–µ–¥–ª–æ–∂–µ–Ω–∏–π")
            print("   2. –ü—Ä–æ–≤–∞–π–¥–µ—Ä —Ä–∞–∑–±–∏–≤–∞–µ—Ç –Ω–∞ —á–∞–Ω–∫–∏ (–≤–Ω—É—Ç—Ä–µ–Ω–Ω—è—è –æ–±—Ä–∞–±–æ—Ç–∫–∞)")
            print("   3. _stream_audio_for_sentence –Ω–∞–∫–∞–ø–ª–∏–≤–∞–µ—Ç –≤—Å–µ —á–∞–Ω–∫–∏")
            print("   4. –ß–∞–Ω–∫–∏ —Å–æ–±–∏—Ä–∞—é—Ç—Å—è –≤ –µ–¥–∏–Ω–æ–µ –∞—É–¥–∏–æ –ø—Ä–µ–¥–ª–æ–∂–µ–Ω–∏—è")
            print("   5. –ü–æ–ª–Ω–æ–µ –∞—É–¥–∏–æ –æ—Ç–ø—Ä–∞–≤–ª—è–µ—Ç—Å—è —á–µ—Ä–µ–∑ gRPC –∫–ª–∏–µ–Ω—Ç—É")
            print("   6. –ö–∞–∂–¥–æ–µ –ø—Ä–µ–¥–ª–æ–∂–µ–Ω–∏–µ –∏–º–µ–µ—Ç –æ–¥–Ω–æ –ø–æ–ª–Ω–æ–µ –∞—É–¥–∏–æ (–Ω–µ –æ—Ç–¥–µ–ª—å–Ω—ã–µ —á–∞–Ω–∫–∏)")
            print("   7. –¢–µ–∫—Å—Ç –∏ –∞—É–¥–∏–æ —Å–æ–æ—Ç–≤–µ—Ç—Å—Ç–≤—É—é—Ç –ø–æ sentence_index")
            return True
        else:
            print("\n" + "="*80)
            print("‚ùå –ù–ï–ö–û–¢–û–†–´–ï –ü–†–û–í–ï–†–ö–ò –ü–†–û–í–ê–õ–ï–ù–´")
            print("="*80)
            return False
        
    except Exception as e:
        print(f"\n‚ùå –û–®–ò–ë–ö–ê: {e}")
        import traceback
        traceback.print_exc()
        return False
    finally:
        # Cleanup
        if 'audio_processor' in locals():
            await audio_processor.cleanup()


async def main():
    """–ó–∞–ø—É—Å–∫ —Ç–µ—Å—Ç–∞"""
    print("="*80)
    print("–¢–ï–°–¢ –ü–û–õ–ù–û–ô PIPELINE: –ì–ï–ù–ï–†–ê–¶–ò–Ø ‚Üí –ù–ê–ö–û–ü–õ–ï–ù–ò–ï ‚Üí –°–ë–û–†–ö–ê ‚Üí –ü–ï–†–ï–î–ê–ß–ê")
    print("="*80)
    
    success = await test_chunk_collection_pipeline()
    
    if success:
        print("\nüéâ –¢–ï–°–¢ –ü–†–û–ô–î–ï–ù –£–°–ü–ï–®–ù–û!")
        return 0
    else:
        print("\n‚ö†Ô∏è  –¢–ï–°–¢ –ü–†–û–í–ê–õ–ï–ù")
        return 1


if __name__ == "__main__":
    exit_code = asyncio.run(main())
    sys.exit(exit_code)
